timemachine/md/builders.py

```diff
@@ -77,24 +92,83 @@ def get_waters_to_delete():
         waters_to_delete = set()
         for idx in clashy_idxs:
             atom = all_atoms[idx]
-            waters_to_delete.add(atom.residue)
-            assert atom.residue.name == WATER_RESIDUE_NAME
+            if atom.residue.name == WATER_RESIDUE_NAME:
+                waters_to_delete.add(atom.residue)
         return waters_to_delete
 
     # First add back in the number of waters that are clashy. Then delete the clashy waters.
     # Done in this order so that additional waters being added will be at the boundaries, if added after deleting
     # addSolvent fills the void intended for the mols. Need to end up with the same number of waters as originally
     num_system_atoms = host_coords.shape[0]
     clashy_waters = get_waters_to_delete()
+    combined_templates = get_ion_residue_templates(modeller)
     # First add back in the number of waters that are clashy and we know we need to delete
-    modeller.addSolvent(host_ff, numAdded=len(clashy_waters), neutralize=False, model=sanitize_water_ff(water_ff))
+    modeller.addSolvent(
+        host_ff,
+        numAdded=len(clashy_waters),
+        neutralize=False,
+        model=sanitize_water_ff(water_ff),
+        residueTemplates=combined_templates,
+    )
     clashy_waters = get_waters_to_delete()
     modeller.delete(list(clashy_waters))
     assert num_system_atoms == modeller.getTopology().getNumAtoms()
 
 
+def solvate_modeller(
+    modeller: app.Modeller,
+    box: NDArray[np.float64],
+    ff: app.ForceField,
+    water_ff: str,
+    mols: list[Chem.Mol] | None,
+    ionic_concentration: float = 0.0,
+    neutralize: bool = False,
+):
+    """Utility function to handle the neutralization of adding mols"""
```

Comment:
Maybe also note that this is a non-idempotent method that will directly modify the modeler object? (Or maybe it should deep copy and return a clone?)